lib/pester.rb

```diff
@@ -0,0 +1,96 @@
+  #
+  # Options:
+  #   retry_error_classes       - A single or array of exceptions to retry on. Thrown exceptions not in this list
+  #                               (including parent/sub-classes) will be reraised
+  #   reraise_error_classes     - A single or array of exceptions to always re-raiseon. Thrown exceptions not in
+  #                               this list (including parent/sub-classes) will be retried
+  #   max_attempts              - Max number of attempts to retry
+  #   delay_interval            - Second interval by which successive attempts will be incremented. A value of 2
+  #                               passed to retry_with_backoff will retry first after 2 seconds, then 4, then 6, et al.
+  #   on_retry                  - A Proc to be called on each successive failure, before the next retry
+  #   on_max_attempts_exceeded  - A Proc to be called when attempt_num >= max_attempts - 1
+  #   message                   - String or regex to look for in thrown exception messages. Matches will trigger retry
+  #                               logic, non-matches will cause the exception to be reraised
+  #
+  # Usage:
+  #   retry_action do
+  #     puts 'trying to remove a directory'
+  #     FileUtils.rm_r(directory)
+  #   end
+  #
+  #  retryable(error_classes: Mysql2::Error, message: /^Lost connection to MySQL server/, max_attempts: 2) do
+  #    ActiveRecord::Base.connection.execute("LONG MYSQL STATEMENT")
+  #  end
+  def self.retry_action(opts = {}, &block)
+    merge_defaults(opts)
+    if opts[:retry_error_classes] && opts[:reraise_error_classes]
+      fail 'You can only have one of retry_error_classes or reraise_error_classes'
+    end
+
+    opts[:max_attempts].times do |attempt_num|
+      begin
+        result = yield block
+        return result
+      rescue => e
+        class_reraise = opts[:retry_error_classes] && !opts[:retry_error_classes].include?(e.class)
+        reraise_error = opts[:reraise_error_classes] && opts[:reraise_error_classes].include?(e.class)
+        message_reraise = opts[:message] && !e.message[opts[:message]]
+
+        if class_reraise || message_reraise || reraise_error
+          match_type = class_reraise ? 'class' : 'message'
+          opts[:logger].warn("Reraising exception from inside retry_action because provided #{match_type} was not matched.")
+          raise
+        end
+
+        if opts[:max_attempts] - 1 > attempt_num
+          attempts_left = opts[:max_attempts] - attempt_num - 1
+          trace = e.backtrace
+          opts[:logger].warn("Failure encountered: #{e}, backing off and trying again #{attempts_left} more times. Trace: #{trace}")
+          opts[:on_retry].call(attempt_num, opts[:delay_interval])
+        else
+          return opts[:on_max_attempts_exceeded].call(opts[:logger], opts[:max_attempts], e)
+        end
+      end
+    end
+  end
+
+  private
+
+  def self.logger
+    if defined? Rails
+      Rails.logger
+    else
+      require 'logger'
```

Comment:
huh maybe i'm just too used to running rails instead of ruby

On Sat, Mar 28, 2015 at 5:11 PM, Marc Bollinger <notifications@github.com>
wrote:

> In lib/pester.rb
> <https://github.com/lumoslabs/pester/pull/1#discussion_r27349992>:
>
> > +          opts[:logger].warn("Failure encountered: #{e}, backing off and trying again #{attempts_left} more times. Trace: #{trace}")
> > +          opts[:on_retry].call(attempt_num, opts[:delay_interval])
> > +        else
> > +          return opts[:on_max_attempts_exceeded].call(opts[:logger], opts[:max_attempts], e)
> > +        end
> > +      end
> > +    end
> > +  end
> > +
> > +  private
> > +
> > +  def self.logger
> > +    if defined? Rails
> > +      Rails.logger
> > +    else
> > +      require 'logger'
>
> Negative, definitely fails. Not a ruby guy, but my guess is that out of
> StdLib <http://ruby-doc.org/stdlib-2.2.1/packages.html>, only the core is
> included by default. I'm familiar with requiring like date and such in
> irb, so I'm guessing non-core packages just need to be required somewhere
> along the line (e.g. magically by rails or whatever).
>
> marc-mbp-2:~ marc$ irb
> irb(main):001:0> Logger
> NameError: uninitialized constant Logger
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/lumoslabs/pester/pull/1/files#r27349992>.
>